### ISSUE TITLE
If two or more "location" block defines the same "js_var" directive name, the value will be overwritten

### ISSUE THREAD

**Issue by HorseLuke** [nginx/njs#667]
**Abstract:**

If two or more "location" block defines the same "js_var" directive name, the value in the previous "location" block will be overwritten by the next "location" block. This behaviour is not the same as the "set" directive, it should be consider as a bug.

---

**Detail:**

js_content directive can pass "module.function" which only accept one paramter `(r: NginxHTTPRequest)`, but sometimes developer want to pass more paramters by location block with same "module.function". js_var or set may be used as this idea. 

The problem is, when use the same name in js_var, the last one will overwrite the previous one. that means value isolation by location block fail.

This behavior does not exist in the "set" directive, so it should be consider as a bug.

---

**Environment & Version:**

CentOS 7
nginx.x86_64 1:1.25.2-1.el7.ngx
nginx-module-njs.x86_64 1:1.25.2+0.8.0-1.el7.ngx

---

**Steps to reproduce:**

nginx.conf:

```
server {
    listen       8765;
    server_name  localhost;
    

    js_import test from test.js;

    location /test {
        js_var $js_var_test "js_var_test from /test";
        set $set_test "set_test from /test";
        js_content test.test;
    }

    location /test2 {
        js_var $js_var_test "js_var_test from /test2";
        set $set_test "set_test from /test2";
        js_content test.test;
    }

}

```

test.js:

```
function test(r){
    r.return(200, JSON.stringify({
        "js_var_test": r.variables.js_var_test,
        "set_test": r.variables.set_test
    }));
}


export default {test};

```

curl /test:

`curl -v "http://localhost:8765/test"
`
curl /test2:

`curl -v "http://localhost:8765/test2"
`

---

**What is expected?**

Started from "from" string, the value from js_var_test should be the same as set_test.

---

**What is actually happening?**

curl /test:

`{"js_var_test":"js_var_test from /test2","set_test":"set_test from /test"}
`
(js_var_test is overwritten)

curl /test2:

`{"js_var_test":"js_var_test from /test2","set_test":"set_test from /test2"}
`

**Comment by xeioex** [comment]
Hi @HorseLuke,

Thank you for your report.

[js_set](http://nginx.org/en/docs/http/ngx_http_js_module.html#js_var) is specifically implemented in a different way to[ set](http://nginx.org/en/docs/http/ngx_http_rewrite_module.html#set). If they are the same we do not need both.

They are similar and the main difference in that js_set is not overwritten after r.internalRedirect() to a new location.
This is exactly what you are seeing with set variables.

For example:
```
    location /test {
        # when you hit /test location the value for $set_test is calculated
        # in a rewrite phase, every time your request hits the location regardless
        # of njs
        set $set_test "AAA";
    }

    location /test2 {
        # when you hit /test location the value for $set_test is calculated
        # in a rewrite phase, every time your request hits the location regardless
        # of njs
        set $set_test "BBB";
    }
    
    # both $set_test are the same variable, but populated by different init code depending on a location
        
```

`js_var` on the other hand is never evaluated until it is first accessed by njs or other directives.
NJS looks for a variable by name using [ngx_http_get_variable](https://github.com/nginx/nginx/blob/master/src/http/ngx_http_variables.c#L689). Which looks for a variable by name in a global variable hash.

In summary: I cannot use the same init mechanism that works on rewrite phase which `set`  is using, because it will break the indented semantic with internal redirects. Without it I do not see a good solution for location matching during a variable evaluation.

**Comment by HorseLuke** [comment]
Sorry for delayed response.

So if I want to use same  "module.function" with  js_content directive in difference location block, how to pass difference arguments for "module.function"?

**Comment by xeioex** [comment]
Hi @HorseLuke,

Please, elaborate more on what you are trying to achieve. Does `set` solve the problem or not?

**Comment by HorseLuke** [comment]
Now I use global ```set``` to solve my problem, the solution is added in https://github.com/nginx/njs/issues/668#issuecomment-1720345674 .

But I'd like to share why I opened those 2 issues.

My scenario is simple, use njs as a auth gateway to auth client token, if auth success then pass to backend with authencated userinfo, if not then show custom error to client.

You may wonder why not use ```auth_request```(then use ```js_content``` in it) , because:

(1) if ```auth_request``` fail, it can not output custom error info to client by fail condition. For example, if token expired then output "token_exipred" , if user is banned then output "user_banned". Now ```auth_request``` only return 401 httpcode without custom info, and can not pass error info to 401 httpcode location.

(2) if ```auth_request``` success, pass success info (e.g., authencated userinfo) to upstream is not in natural way. The only way for ```js_content``` in ```auth_request``` is to use ```headerOut```.

If not use ```auth_request```, another way is to use named location like this:

nginx.conf

```
server {

  set $param_js_content_param_named_location "@not_exists";

  js_var $data_from_js_content_auth_user_info "";

  js_import main from ../dist/main.js;

    //..omit...


  location = /testapp/v1/get{
    set $param_js_content_param_named_location "@upstream_testapp1";
    js_content main.doAuthAndRedirect;
  }

  location = /testapp2/v1/get{
    set $param_js_content_param_named_location "@upstream_testapp2";
    js_content main.doAuthAndRedirect;
  }

  location @upstream_testapp1{
     proxy_pass  http://127.0.0.1:6871;
     proxy_set_header X-Authed-User-Info $data_from_js_content_auth_user_info;
    //..omit...
  }

  location @upstream_testapp2{
     proxy_pass  http://127.0.0.1:7892;
     proxy_set_header X-Authed-User-Info $data_from_js_content_auth_user_info;
    //..omit...

  }

  location @not_exists{
    //..omit...
  }

}
```

(If auth success, js_content will write auth userinfo to ```r.variables["data_from_js_content_auth_user_info"]```, then use ```r.internalRedirect(r.variables["param_js_content_param_named_location"])```)

You can see that every location has the same ```js_content``` with difference param, which trying to solve ```js_content``` can not pass extra param problem. But at the first time I encounter problem no matter I use ```js_var``` or ```set``` , and the behavior is like bugs,  so I opened these 2 issues. Now it seems it is by-design, I find a way to pass it.

**Comment by HorseLuke** [comment]
Closed as "by-design"

**Comment by lcrilly** [comment]
> (1) if auth_request fail, it can not output custom error info to client by fail condition. For example, if token expired then output "token_exipred" , if user is banned then output "user_banned". Now auth_request only return 401 httpcode without custom info, and can not pass error info to 401 httpcode location.

Not true. [auth_request_set](https://nginx.org/en/docs/http/ngx_http_auth_request_module.html#auth_request_set) can be used to return metadata from an auth_request/js_content location that can then be used as part of an `error_page` location.

Does that help?

**Comment by lcrilly** [comment]
> auth_request can not do that

It *can* when combined with `error_page`. With `js_content` you are in control of the response code from `auth_request` and so you can return any response you wish as a result of the code, headers or body of the subrequest.

**Comment by drsm** [comment]
@HorseLuke 

Take a look , https://github.com/nginx/njs/issues/362#issuecomment-742993025

**Comment by HorseLuke** [comment]
> @HorseLuke
> 
> Take a look , [#362 (comment)](https://github.com/nginx/njs/issues/362#issuecomment-742993025)

Thanks for sharing. In fact that is the first version I implemented. 
But as what I said above, passing variables across block directive will add more complexity of the code, and iti is very difficult to maintain when in production env. When encounter bugs, you have to jump between nginx config file and njs module, teammates will not happy and hate it. So that is not a straight solution.

**Comment by HorseLuke** [comment]
> > auth_request can not do that
> 
> It _can_ when combined with `error_page`. With `js_content` you are in control of the response code from `auth_request` and so you can return any response you wish as a result of the code, headers or body of the subrequest.

Emmm......Maybe we are not in the same topic. 

My topic is to take full control in njs module as much as possible, and not to jump across config file and njs module. This will make test easier, and reuse js code easier.

What you said is to split functions into config blocks. I tried it, and my conclusion is that it add complexity and not easy to test and maintain.

**Comment by lcrilly** [comment]
I understand your goal. In general, it's better to use native nginx functionality when it's available. It does add complexity. Some of that can be helped with include files and clear naming of functions and locations, but you get better performance as a result. It's a tradeoff.

**Comment by lcrilly** [comment]
@salacr please try using `=` with the `error_page` directive with one of these:

```nginx
error_page 500 =503 @process_error_500; # Override the status code directly
error_page 500 = @process_error_500;    # Delegate setting the status code to the named location
```

https://nginx.org/en/docs/http/ngx_http_core_module.html#error_page

**Comment by salacr** [comment]
That's it!! Thanks!
